app/validators/projects_validator.rb

```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class ProjectsValidator < ActiveModel::Validator
+  def validate(digital_object)
+    digital_object.errors[:projects] << "Only primary projects can be assigned to primary_project" if digital_object.primary_project && !digital_object.primary_project.is_primary
```

Comment:
Raising an exception in the primary_project/other_projects setter code will prevent unintentional `nil.is_primary` calls here, so that's good.